dali/c_api_2/pipeline_test.cc

```diff
@@ -0,0 +1,508 @@
+  Pipeline p(max_batch_size, num_threads, device_id);
+  OpSpec op1 = TestOp("op1", "cpu")
+    .AddInput("ctr", StorageDevice::CPU)
+    .AddOutput("op1", StorageDevice::CPU)
+    .AddArg("addend", 1000);
+
+  OpSpec op2 = TestOp("op2", "gpu")
+    .AddInput("ctr", StorageDevice::GPU)
+    .AddOutput("op2", StorageDevice::GPU)
+    .AddArg("addend", 2000);
+
+  p.AddOperator(CounterOp("ctr"), "ctr");
+  p.AddOperator(op1, "op1");
+  p.AddOperator(op2, "op2");
+  p.SetOutputDescs({ {"op1", "cpu" }, {"op2", "gpu"} });
+
+  return p.SerializeToProtobuf();
+}
+
+std::string GetGPU2CPUPipelineProto(int max_batch_size, int num_threads, int device_id) {
+  Pipeline p(max_batch_size, num_threads, device_id);
+  OpSpec op1 = TestOp("op1", "cpu")
+    .AddInput("ctr", StorageDevice::CPU)
+    .AddOutput("op1", StorageDevice::CPU)
+    .AddArg("addend", 1000);
+
+  OpSpec op2 = TestOp("op2", "gpu")
+    .AddInput("ctr", StorageDevice::GPU)
+    .AddOutput("op2", StorageDevice::GPU)
+    .AddArg("addend", 2000);
+
+  p.AddOperator(CounterOp("ctr"), "ctr");
+  p.AddOperator(op1, "op1");
+  p.AddOperator(op2, "op2");
+  p.SetOutputDescs({ {"op1", "cpu" }, {"op2", "cpu"} });
+
+  return p.SerializeToProtobuf();
+}
+
+std::string GetPipelineWithTraces(int max_batch_size, int num_threads, int device_id) {
+  Pipeline p(max_batch_size, num_threads, device_id);
+  OpSpec op1 = TraceOp("op1", "cpu", "t1")
+    .AddInput("ctr", StorageDevice::CPU)
+    .AddOutput("op1", StorageDevice::CPU);
+
+  OpSpec op2 = TraceOp("op1", "gpu", "t2")
+    .AddInput("ctr", StorageDevice::GPU)
+    .AddOutput("op2", StorageDevice::GPU);
+
+  p.AddOperator(CounterOp("ctr"), "ctr");
+  p.AddOperator(op1, "op1");
+  p.AddOperator(op2, "op2");
+  p.SetOutputDescs({ {"op1", "cpu" }, {"op2", "cpu"} });
+
+  return p.SerializeToProtobuf();
+}
+
+std::string GetPipelineWithExternalSource(
+      StorageDevice dev_type,
+      int max_batch_size,
+      int num_threads,
+      int device_id) {
+  Pipeline p(max_batch_size, num_threads, device_id);
+  OpSpec src = OpSpec("ExternalSource")
+    .AddOutput("out", dev_type)
+    .AddArg("device", dev_type == StorageDevice::CPU ? "cpu" : "gpu")
+    .AddArg("name", "ext")
+    .AddArg("batch_size", max_batch_size);
+  p.AddOperator(src, "ext");
+  p.SetOutputDescs({ {"out", to_string(dev_type)} });
+  return p.SerializeToProtobuf();
+}
+
+namespace c_api {
+namespace test {
+
+inline Pipeline *GetPipeline(daliPipeline_h h) {
+  return h ? static_cast<PipelineWrapper *>(h)->Unwrap() : nullptr;
+}
+
+TEST(CAPI2_PipelineTest, Deserialize) {
+  std::string proto = GetCPUOnlyPipelineProto(4, 4, 0);
+
+  daliPipelineParams_t params{};
+  daliPipeline_h h = nullptr;
+  CHECK_DALI(daliPipelineDeserialize(&h, proto.c_str(), proto.length(), &params));
+  ASSERT_NE(h, nullptr);
+  CHECK_DALI(daliPipelineBuild(h));
+  EXPECT_EQ(GetPipeline(h)->output_descs().size(), 2);
+  CHECK_DALI(daliPipelineDestroy(h));
```

Comment:
Shouldn't other test call this as well? I don't see it in other cases.

Reply:
No, other tests use a managed handle. It's safer should an exception be thrown - and re-testing the return value of `daliPipelineDestroy` doesn't bring much value, since the only error it can ever produce is invalid handle - an exception in the pipeline destructor will result in program termination.